### PR TITLE
Custom Salt States

### PR DESCRIPTION
**Pull request by Loquacity**
# Description

We mention custom states but don't actually say how to do it. This PR adds a new section for custom states, provides an example, and includes a custom state for trusting GPG keys (to satisfy https://github.com/SUSE/spacewalk/issues/13812). It also updates some of the other references to custom states.

# Target branches

Which documentation version does this PR apply to?

- [x] Master (Default)
- [ ] Manager-4.1
- [ ] Manager-4.0

# Links

- Fixes https://github.com/SUSE/spacewalk/issues/13812
- Also ref: https://github.com/uyuni-project/uyuni/issues/3133

### CODE REVIEW

File: modules/salt/pages/custom-states.adoc [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+You can create your own custom Salt states with {productname} as centrally managed configuration channels.
+Custom states are stored as Salt state files on the {productname} Server with a ``.sls`` extension, in the [path]``/srv/susemanager/salt/`` directory.
+
+
+
+== Create a New Custom Salt Channel
+
+You can use the {productname} {webui} to create and edit custom Salt state files.
+You must create a state channel first, with an initial state named ``init.sls``.
```
Comment (Bischoff) [inline modules/salt/pages/custom-states.adoc:12]: As a complete noob, I understand this as 
```
you must create the state channel first, and then you need to
create the `init.sls` file
```
but apparently this states file will be created automatically when you create the state channel.

If that is so, I would rephrase last sentence as.
```
You must create a state channel first. This newly created channel
will contain an initial states file named ``init.sls``.
```

@juliogonzalez can you please confirm I'm not saying crap, here?

File: modules/salt/pages/custom-states.adoc [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  For example, ``GPG_Key_Trusts``
+. In the [guimenu]``Description`` field, type a short description of the configuration your state performs.
+  For example, ``Trusts GPG Keys for CentOS``.
+. In the [guimenu]``SLS Contents`` field, create a state to retrieve the appropriate key from the {productname} Server and trust it on the client.
+  For example FIXME:
++
+----
+cmd.run:
+  - salt '*' gpg.import_key filename='/srv/www/htdocs/pub/centos8-gpg-pubkey-05B555B38483C65D.key'
+  - salt '*' gpg.trust_key keyid='centos8-gpg-pubkey-05B555B38483C65D.key' trust_level='trusted'
```
Comment (Bischoff) [inline modules/salt/pages/custom-states.adoc:142]: Using Salt to call Salt seems a bit wird to me... Isn't there something more elegant and straightforward?

**Review by Bischoff — CHANGES_REQUESTED** [review]
Very good start, thanks a lot. See my 2 suggestions for changes.

File: modules/salt/pages/custom-states.adoc (hunk 1 above)
Comment (Loquacity) [reply modules/salt/pages/custom-states.adoc:12]: When I went through the process, it's kind of not automatically created, though. I need to type something in the box.

File: modules/salt/pages/custom-states.adoc (hunk 2 above)
Comment (Loquacity) [reply modules/salt/pages/custom-states.adoc:142]: This was a wild guess on my part. I need someone who knows what they're doing here :stuck_out_tongue:

Comment (Loquacity) [reply modules/salt/pages/custom-states.adoc:142]: Perhaps @juliogonzalez can help?

File: modules/salt/pages/custom-states.adoc (hunk 1 above)
Comment (Bischoff) [reply modules/salt/pages/custom-states.adoc:12]: thanks for the explanation Lana

File: modules/salt/pages/custom-states.adoc (hunk 2 above)
Comment (Loquacity) [reply modules/salt/pages/custom-states.adoc:142]: @juliogonzalez if you can't help, or don't have bandwidth, can you recommend someone else who could?

Comment (mcalmer) [reply modules/salt/pages/custom-states.adoc:142]: I would need to check, but i fear this would not do what you expect.

Comment (Loquacity) [reply modules/salt/pages/custom-states.adoc:142]: No, I wasn't really expecting that it would. Who could tell me?

File: modules/salt/pages/custom-states.adoc [hunk 3]
```diff
@@ -0,0 +1,151 @@
+[[custom-states]]
+= Custom Salt States
+
+You can create your own custom Salt states with {productname} as centrally managed configuration channels.
+Custom states are stored as Salt state files on the {productname} Server with a ``.sls`` extension, in the [path]``/srv/susemanager/salt/`` directory.
```
Comment (mcalmer) [inline modules/salt/pages/custom-states.adoc:5]: `/srv/susemanager/salt/` has content which is managed by SUMA. I think no customer should mess around with files in this directory. So I am not sure if it is wise to tell them where it is. At least not at such a prominent place?

File: modules/salt/pages/custom-states.adoc [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+. In the [guimenu]``Label`` field, type a label.
+  Use alphanumeric characters, hyphens, and underscores.
+  Do not use spaces.
+. In the [guimenu]``Description`` field, type a short description of the configuration your state performs.
+. In the [guimenu]``SLS Contents`` field, type the contents of your ``init.sls`` state.
+  Ensure your file starts by specifying the source of the managed file, using this syntax:
++
+----
+file.managed:
+  - source: salt://<org_name>/<channel_name>/etc/<filename>
```
Comment (mcalmer) [inline modules/salt/pages/custom-states.adoc:34]: There is always an "ID" required and in most of the cases also a "name".

File: modules/salt/pages/custom-states.adoc [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+
+.Example: Initial State File
+
+----
+file.managed:
+  - source: salt://example_org/example_channel/etc/motd
+[
+  FIXME: reference other file here somehow
+]
```
Comment (mcalmer) [inline modules/salt/pages/custom-states.adoc:84]: ID and name is missing

File: modules/salt/pages/custom-states.adoc [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+update packages:
+  module.run:
+    - name: pkg.upgrade
+    - failhard: True
+check if reboot is needed:
+  salt.set_reboot_needed:
+    - failhard: True
+reboot:
+   salt.reboot_if_needed:
+     - failhard: True
```
Comment (mcalmer) [inline modules/salt/pages/custom-states.adoc:106]: where does this example come from? Is it how we want our customers to do things?

File: modules/salt/pages/custom-states.adoc [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+----
+
+
+
+== Custom State to Trust a GPG Key
+
+By default, operating systems trust only their own GPG keys when they are installed, and do not trust keys provided by third party packages.
+The clients can be successfully bootstrapped without the GPG key being trusted.
+However, you cannot install new client tool packages or update them until the keys are trusted.
```
Comment (mcalmer) [inline modules/salt/pages/custom-states.adoc:115]:
```suggestion
However, you cannot install new third party packages or update them until the keys are trusted.
```

File: modules/salt/pages/custom-states.adoc [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+Salt clients are set to trust {suse} tools channels GPG keys when they are bootstrapped.
+For all other clients and channels, you need to manually trust third party GPG keys.
```
Comment (mcalmer) [inline modules/salt/pages/custom-states.adoc:118]: I would suggest from now on we talk about 3rd party channels and packages. We better use not examples with SLES or SUSE Manager keys prevent the assumption that other SUSE provided keys must be trusted on the clients.

File: modules/salt/pages/custom-states.adoc [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+If you are bootstrapping Salt clients from the {productname} {webui}, you can use a custom Salt state to trust the GPG key.
+
+
+
+.Procedure: Trusting a GPG Key With a Custom Salt State
+. On the {productname} Server, at the command prompt, check the contents of the [path]``/srv/www/htdocs/pub/`` directory.
+  This directory contains all available public keys.
+  Take a note of the key that applies to the channel assigned to the client you are registering.
```
Comment (mcalmer) [inline modules/salt/pages/custom-states.adoc:127]: This directory contains keys we ship, but in most cases these are not the keys the customer has to trust manually.
We should concentrate on cases where a complete new key is needed.

1. find the key which needs to be trusted. Check the web page of the vendor or ask a key server when the keyid is known. Check the fingerprint to verify, that the correct key is used.
2. store the key at a place where client can download it. For example `/srv/www/htdocs/pub/` where we also store our keys

... continued below

File: modules/salt/pages/custom-states.adoc (hunk 2 above)
Comment (mcalmer) [reply modules/salt/pages/custom-states.adoc:142]: Problem: It would be better if we would have a salt module for trusting GPG keys on different OSes, but this does not exist yet. So trusting GPG keys depends on the OS!

```
rpm_trust_gpg_key:
  cmd.run:
    - name: rpm --import https://{{ salt['pillar.get']('mgr_server') }}/pub/third-party-gpg.key
    - unless: rpm -q gpg-pubkey-<key_id>

deb_trust_gpg_key:
  mgrcompat.module_run:
    - name: pkg.add_repo_key
    - path: https://{{ salt['pillar.get']('mgr_server') }}/pub/third-party-gpg.key
```

Comment (mcalmer) [reply modules/salt/pages/custom-states.adoc:142]: Of cause you can also add the GPG key files to the config channel, use a `file.managed` to deploy them on the minion and use a local path instead of a `https://` URL .

File: modules/salt/pages/custom-states.adoc (hunk 6 above)
Comment (Loquacity) [reply modules/salt/pages/custom-states.adoc:106]: One of the Aussie devs provided this to me when I couldn't find one, and I edited the hell out of it. I have no idea what we want to tell customers, this is the best I could do with what I had to work with. Any/all suggestions welcome!

**Review by keichwa — APPROVED** [review]

File: modules/salt/pages/custom-states.adoc [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+
+
+== Example Custom State Files
+
+This section contains some example custom state files.
+Use these as a basis for writing your own custom states.
+
+
+
+.Example: Initial State File
```
Comment (mcalmer) [inline modules/salt/pages/custom-states.adoc:77]:
```suggestion
.Example: Manage a File
```

Initial State File is just a state with the name of the channel. It has no special content.
It can contain whatever the customer want to manage. There is no need to manage additional files at all.

I would suggest to list here a number of the mostly used state modules.

- file.managed
- pkg.installed
- cmd.run
- service.running

File: modules/salt/pages/custom-states.adoc [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+
+----
+file.managed:
+  - source: salt://example_org/example_channel/etc/ID/filename
+[
+  FIXME: reference other file here somehow
+]
+----
```
Comment (mcalmer) [inline modules/salt/pages/custom-states.adoc:85]:
```suggestion
----
my_config_change_id:
  file.managed:
    -name: /etc/my.conf
    - source: salt://example_org/example_channel/etc/my.conf
    - user: root
    - group: root
    - mode: 644
    - template: jinja
----
```

File: modules/salt/pages/custom-states.adoc [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+
+
```
Comment (mcalmer) [inline modules/salt/pages/custom-states.adoc:87]:
```suggestion
.Example: Package Management

----
my_pkg_id:
  pkg.installed:
    - refresh: True
    - pkgs:
      - glibc
      - kernel-default
      - hello: 1.0-42
----

.Example: Remote Command

----
ip_forward-on:
  cmd.run:
    - name: echo "1" > /proc/sys/net/ipv4/ip_forward
    - onlyif:
      - test `cat /proc/sys/net/ipv4/ip_forward` -eq 0
----

.Example: Service Management

----
time_service_id:
  service.running:
    - name: chronyd
    - enable: True
----
```

**Review by mcalmer — COMMENTED** [review]
Left some comments

File: modules/salt/pages/custom-states.adoc (hunk 6 above)
Comment (mcalmer) [reply modules/salt/pages/custom-states.adoc:141]: Please remove this example. At least the last states `salt.set_reboot_needed` and `salt.reboot_if_needed` seems to be unavailable. They do not exist.

A state keeping things up to date is nice, but hard to write. We better do not describe something which is not working.

**Review by mcalmer — CHANGES_REQUESTED** [review]
Only 1 problem left

**Review by mcalmer — APPROVED** [review]

**Comment by Loquacity** [comment]
Thanks @mcalmer :sparkling_heart:

File: modules/salt/pages/custom-states.adoc [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
++
+ Example custom state files are given later in this section.
+. Click btn:[Update Channel] to save your state.
+
+
+.Procedure: Adding Additional Files to a Custom State Channel
+. In the {productname} {webui}, navigate to menu:Configuration[Channels].
+. Click the name of the channel you want to add files to.
+. To create a new file, click btn:``Create configuration file`` and type the contents of the file.
+. To upload an existing file, click btn:``Upload Configuration Files`` and select the file to upload.
```
Comment (keichwa) [inline modules/salt/pages/custom-states.adoc:45]:
```suggestion
. To upload an existing file, click btn:[Upload Configuration Files] and select the file to upload.
```

File: modules/salt/pages/custom-states.adoc [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+. To copy an existing file, click btn:``Import a File from Another Channel or System`` and select the file to copy.
```
Comment (keichwa) [inline modules/salt/pages/custom-states.adoc:46]:
```suggestion
. To copy an existing file, click btn:[Import a File from Another Channel or System] and select the file to copy.
```

File: modules/salt/pages/custom-states.adoc [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+
+
+
+.Procedure: Editing a Custom Salt State
+. In the {productname} {webui}, navigate to menu:Configuration[Channels].
+. Click btn:[View/Edit <filename>.sls File].
+. Make your changes to the file
```
Comment (keichwa) [inline modules/salt/pages/custom-states.adoc:53]:
```suggestion
. Make your changes to the file.
```

File: modules/salt/pages/custom-states.adoc [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+
+.Procedure: Assigning a Client to a Custom Salt State
+. In the {productname} {webui}, navigate to menu:Configuration[Channels].
+. Click the name of the state you want to assign a client to.
+. Navigate to the menu:Systems[Target Systems] tab.
+. Check the clients you want to assign.
+. Click btn:[Subscribe systems].
+
+
+For more information about Salt state modules, see https://docs.saltproject.io/en/latest/ref/states/all/index.html[].
```
Comment (keichwa) [inline modules/salt/pages/custom-states.adoc:68]:
```suggestion
For more information about Salt state modules, see https://docs.saltproject.io/en/latest/ref/states/all/index.html.
```

File: modules/salt/pages/custom-states.adoc [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+Use these as a basis for writing your own custom states.
+
+
+
+.Example: Manage a File
+
+----
+my_config_change_id:
+  file.managed:
+    -name: /etc/my.conf
```
Comment (keichwa) [inline modules/salt/pages/custom-states.adoc:84]:
```suggestion
    - name: /etc/my.conf
```

File: modules/salt/pages/custom-states.adoc [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+.Procedure: Trusting a GPG Key With a Custom Salt State
+. Locate the key that you need to trust.
+  Ensure you have the correct key, and that you also have the fingerprint used to verify the key.
+  This information is available from the vendor or, in some cases, from a key server.
+. Copy the key to a file location where the client can access it.
+  We recommend saving it in the [path]``/srv/www/htdocs/pub/`` directory, where all {suse} public keys are also saved.
+. In the {productname} {webui}, navigate to menu:Configuration[Channels].
+. Click btn:[Create State Channel].
+. In the [guimenu]``Name`` field, type a name for your state.
+  For example, ``GPG Key Trusts``
```
Comment (keichwa) [inline modules/salt/pages/custom-states.adoc:150]:
```suggestion
  For example, ``GPG Key Trusts``.
```

File: modules/salt/pages/custom-states.adoc [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+. In the [guimenu]``Label`` field, type a label.
+  For example, ``GPG_Key_Trusts``
```
Comment (keichwa) [inline modules/salt/pages/custom-states.adoc:152]:
```suggestion
  For example, ``GPG_Key_Trusts``.
```

**Review by keichwa — APPROVED** [review]

**Review by keichwa — APPROVED** [review]
